src/bun.js/test/snapshot.zig

```diff
@@ -0,0 +1,292 @@
+    // maybe use resolve_path.joinAbsString
+    const snapshot_file_path = strings.join(string_list, "", allocator) catch unreachable;
+
+    return fs.Path.init(snapshot_file_path);
+}
+
+fn escapeBackTicks(str: bun.string, allocator: std.mem.Allocator) bun.string {
+    var escaped_string_len: usize = 0;
+    for (str) |v| {
+        if (v == '`')
+            escaped_string_len += 1;
+        escaped_string_len += 1;
+    }
+
+    var i: usize = 0;
+    var buf = allocator.alloc(u8, escaped_string_len) catch unreachable;
+    for (str) |v| {
+        if (v == '`') {
+            buf[i] = '\\';
+            buf[i + 1] = '`';
+            i += 2;
+        } else {
+            buf[i] = v;
+            i += 1;
+        }
+    }
+    return buf;
+}
+
+pub fn createSnapshotName(testName: bun.string, describeScope: *DescribeScope, hint: ZigString, allocator: std.mem.Allocator) bun.string {
+    var arr = std.ArrayList(bun.string).init(allocator);
+    var parent: ?*DescribeScope = describeScope;
+    while (parent) |scope| {
+        // can optimize so that each insert is done in O(1)
+        if (!std.mem.eql(u8, scope.label, "")) arr.insert(0, scope.label) catch unreachable;
+        parent = scope.parent;
+    }
+
+    if (hint.len > 0) {
+        var snap_name_list = [_][]const u8{ testName, ": ", hint.slice() };
+        const snap_name = std.mem.concat(allocator, u8, &snap_name_list) catch unreachable;
+        arr.append(snap_name) catch unreachable;
+        const snapshot_name = strings.join(arr.items, " ", allocator) catch unreachable;
+        return snapshot_name;
+    }
+    arr.append(testName) catch unreachable;
+    const snapshot_name = strings.join(arr.items, " ", allocator) catch unreachable;
+    return snapshot_name;
+}
+
+pub const SnapshotFile = struct {
+    path: fs.Path,
+    globalObject: ?*JSC.JSGlobalObject,
+    allocator: std.mem.Allocator,
+    // This is the data that is read from the snapshot file
+    // data is represented like:
+    // { 'test constructor works': 4 }
+    snapshot_data: SnapshotError!JSC.JSValue,
+    // This is the actual values that the user calls
+    // used to update the snapshot file
+    tests: std.ArrayListUnmanaged(TestData) = .{},
+    counters: CountersMap,
+    update_snapshot: bool = false,
+
+    pub var update_all_snapshots: bool = false;
+
+    pub const TestData = struct {
+        k: ZigString,
+        v: JSC.JSValue,
+        count: u32,
+
+        pub const List = std.ArrayListUnmanaged(TestData);
+
+        fn lessThan(_: void, a: TestData, b: TestData) bool {
+            return ZigString.cmpDesc(void{}, a.k, b.k);
+        }
+    };
+
+    pub const SnapshotError = error{ SnapshotNotFound, CaseNotFound };
+
+    pub const CountersMap = std.HashMap(u64, u32, IdentityContext(u64), 80);
+
+    pub fn readAndParseSnapshot(this: *SnapshotFile, globalObject: *JSC.JSGlobalObject) void {
+        var snapshot_contents = this.readSnapshot() catch return;
```

Comment:
This string is potentially a UTF-8 string that needs to be re-encoded (call `.withEncoding()`)